File: src/main/java/me/testaccount666/serversystem/commands/executables/kit/CommandKit.java
package me.testaccount666.serversystem.commands.executables.kit;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.commands.ServerSystemCommand;
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand;
import me.testaccount666.serversystem.commands.executables.kit.manager.Kit;
import me.testaccount666.serversystem.commands.executables.kit.manager.KitManager;
import me.testaccount666.serversystem.managers.PermissionManager;
import me.testaccount666.serversystem.userdata.ConsoleUser;
import me.testaccount666.serversystem.userdata.User;
import org.bukkit.command.Command;
import org.bukkit.entity.Player;

import static me.testaccount666.serversystem.utils.DurationParser.parseDuration;
import static me.testaccount666.serversystem.utils.DurationParser.parseUnbanDate;
import static me.testaccount666.serversystem.utils.MessageBuilder.command;
import static me.testaccount666.serversystem.utils.MessageBuilder.general;

@ServerSystemCommand(name = "kit", variants = {"createkit", "deletekit"}, tabCompleter = TabCompleterKit.class)
public class CommandKit extends AbstractServerSystemCommand {

    public CommandKit() {
        ServerSystem.Instance.setKitManager(new KitManager());
    }

    @Override
    public boolean hasCommandAccess(Player player, Command command) {
        var permissionPath = getPermission(command);
        return PermissionManager.hasCommandPermission(player, permissionPath, false);
    }

    @Override
    public void execute(User commandSender, Command command, String label, String... arguments) {
        if (arguments.length == 0) {
            general("InvalidArguments", commandSender).label(label).build();
            return;
        }

        if (commandSender instanceof ConsoleUser) {
            general("NotPlayer", commandSender).build();
            return;
        }

        var permissionPath = getPermission(command);
        if (!checkBasePermission(commandSender, permissionPath)) return;

        switch (command.getName().toLowerCase()) {
            case "createkit" -> handleCreateKit(commandSender, label, arguments);
            case "deletekit" -> handleDeleteKit(commandSender, label, arguments);
            case "kit" -> handleKit(commandSender, label, arguments);
        }
    }

    private void handleCreateKit(User commandSender, String label, String... arguments) {
        var kitName = arguments[0].toLowerCase();
        var kitManager = ServerSystem.Instance.getKitManager();
        var kitOptional = kitManager.getKit(kitName);
        if (kitOptional.isPresent()) {
            command("Kit.Create.KitAlreadyExists", commandSender)
                    .postModifier(message -> message.replace("<KIT>", kitOptional.get().getDisplayName())).build();
            return;
        }

        var cooldown = -1L;

        if (arguments.length > 1) cooldown = parseDuration(arguments[1]);
        if (cooldown == -2) {
            command("Kit.Create.InvalidCooldown", commandSender).build();
            return;
        }

        var player = commandSender.getPlayer();
        var inventory = player.getInventory();

        var offHandItem = inventory.getItemInOffHand();
        var armorContents = inventory.getArmorContents();
        var contents = inventory.getContents();

        var kit = new Kit(kitName, cooldown, offHandItem, armorContents, contents);
        kitManager.addKit(kit);
        kitManager.saveAllKits();
        command("Kit.Create.Success", commandSender).postModifier(message -> message.replace("<KIT>", kit.getDisplayName())).build();
    }

    private void handleDeleteKit(User commandSender, String label, String... arguments) {
        var kitName = arguments[0].toLowerCase();
        var kitManager = ServerSystem.Instance.getKitManager();
        if (!kitManager.kitExists(kitName)) {
            command("Kit.KitNotFound", commandSender)
                    .postModifier(message -> message.replace("<KIT>", arguments[0])).build();
            return;
        }

        kitManager.removeKit(kitName);
        command("Kit.Delete.Success", commandSender).postModifier(message -> message.replace("<KIT>", arguments[0])).build();
    }

    private void handleKit(User commandSender, String label, String... arguments) {
        var kitName = arguments[0].toLowerCase();
        var kitManager = ServerSystem.Instance.getKitManager();
        var kitOptional = kitManager.getKit(kitName);
        if (kitOptional.isEmpty()) {
            command("Kit.KitNotFound", commandSender)
                    .postModifier(message -> message.replace("<KIT>", arguments[0])).build();
            return;
        }
        var kit = kitOptional.get();

        var userOptional = getTargetUser(commandSender, 1, arguments);
        if (userOptional.isEmpty()) {
            general("PlayerNotFound", commandSender).target(arguments[1]).build();
            return;
        }
        var targetUser = userOptional.get();
        var targetPlayer = targetUser.getPlayer();
        var isSelf = targetUser == commandSender;

        if (!isSelf && !checkOtherPermission(commandSender, "Kit.Other", targetPlayer.getName())) return;

        if (isSelf && commandSender.isOnKitCooldown(kitName)) {
            var cooldown = commandSender.getKitCooldown(kitName);

            command("Kit.OnCooldown", commandSender)
                    .postModifier(message -> message.replace("<KIT>", kit.getDisplayName())
                            .replace("<DATE>", parseUnbanDate(cooldown))).build();
            return;
        }

        if (isSelf) commandSender.setKitCooldown(kitName, kit.getCoolDown());
        kit.giveKit(targetPlayer);

        var messagePath = "Kit.Success." + (isSelf? "Self" : "Other");
        command(messagePath, commandSender).target(targetPlayer.getName())
                .postModifier(message -> message.replace("<KIT>", kit.getDisplayName())).build();
    }

    private String getPermission(Command command) {
        return switch (command.getName().toLowerCase()) {
            case "createkit" -> "Kit.Create";
            case "deletekit" -> "Kit.Delete";
            case "kit" -> "Kit.Use";
            default -> null;
        };
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/kit/CommandKit.kt
package me.testaccount666.serversystem.commands.executables.kit

import me.testaccount666.serversystem.commands.ServerSystemCommand
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand
import me.testaccount666.serversystem.commands.executables.kit.manager.Kit
import me.testaccount666.serversystem.commands.executables.kit.manager.KitManager
import me.testaccount666.serversystem.extensions.*
import me.testaccount666.serversystem.userdata.User
import me.testaccount666.serversystem.utils.DurationParser.parseDate
import me.testaccount666.serversystem.utils.DurationParser.parseDuration
import org.bukkit.command.Command

@ServerSystemCommand("kit", ["createkit", "deletekit"], TabCompleterKit::class)
class CommandKit : AbstractServerSystemCommand() {
    override fun minRequiredArguments(command: Command) = 1
    override fun getUsagePermission(command: Command): String {
        return when (command.name.lowercase()) {
            "createkit" -> "Kit.Create"
            "deletekit" -> "Kit.Delete"
            "kit" -> "Kit.Use"
            else -> error("(CommandKit;getUsagePermission) Unexpected value: ${command.name}")
        }
    }

    override fun getSyntaxPath(command: Command?): String {
        if (command == null) return "Kit"
        return when (val commandName = command.name.lowercase()) {
            "createkit" -> "CreateKit"
            "deletekit" -> "DeleteKit"
            "kit" -> "Kit"
            else -> error("(CommandKit;SyntaxPath) Unexpected value: ${commandName}")
        }
    }

    override fun execute(commandSender: User, command: Command, label: String, vararg arguments: String) {
        if (!isPlayer(commandSender)) return

        when (command.name.lowercase()) {
            "createkit" -> handleCreateKit(commandSender, *arguments)
            "deletekit" -> handleDeleteKit(commandSender, *arguments)
            "kit" -> handleKit(commandSender, *arguments)
        }
    }

    private fun handleCreateKit(commandSender: User, vararg arguments: String) {
        val kitName = arguments[0]
        val kitManager = getService<KitManager>()
        kitManager.getKit(kitName)?.also { kit ->
            commandSender.commandMsg("Kit.Create.KitAlreadyExists") {
                postModifier { it.replace("<KIT>", kit.displayName) }
            }
            return
        }

        val cooldown = if (arguments.size > 1) parseDuration(arguments[1]) else -1L
        if (cooldown == -2L) {
            commandSender.commandMsg("Kit.Create.InvalidCooldown")
            return
        }

        val inventory = commandSender.getPlayer()!!.inventory

        val newKit = Kit(kitName, cooldown, inventory.contents)
        kitManager.addKit(newKit)
        kitManager.saveAllKits()
        commandSender.commandMsg("Kit.Create.Success") {
            postModifier { it.replace("<KIT>", newKit.displayName) }
        }
    }

    private fun handleDeleteKit(commandSender: User, vararg arguments: String) {
        val kitName = arguments[0]
        val kitManager = getService<KitManager>()
        if (!kitManager.kitExists(kitName)) {
            commandSender.commandMsg("Kit.KitNotFound") {
                postModifier { it.replace("<KIT>", arguments[0]) }
            }
            return
        }

        kitManager.removeKit(kitName)
        commandSender.commandMsg("Kit.Delete.Success") {
            postModifier { it.replace("<KIT>", arguments[0]) }
        }
    }

    private fun handleKit(commandSender: User, vararg arguments: String) {
        val kitName = arguments[0]
        val kitManager = getService<KitManager>()
        val kit = kitManager.getKit(kitName) ?: run {
            commandSender.commandMsg("Kit.KitNotFound") {
                postModifier { it.replace("<KIT>", arguments[0]) }
            }
            return
        }

        val targetUser = getTargetUser(commandSender, 1, arguments = arguments) ?: run {
            commandSender.generalMsg("PlayerNotFound") { target(arguments[1]) }
            return
        }
        val targetPlayer = targetUser.getPlayer()!!
        val isSelf = targetUser === commandSender

        if (!isSelf && !checkPermission(commandSender, "Kit.Other", targetPlayer.name)) return

        if (isSelf && commandSender.isOnKitCooldown(kitName)) {
            val cooldown = commandSender.getKitCooldown(kitName)

            commandSender.commandMsg("Kit.OnCooldown") {
                postModifier {
                    it.replace("<KIT>", kit.displayName)
                        .replace("<DATE>", parseDate(cooldown, commandSender))
                }
            }
            return
        }

        if (isSelf) targetUser.setKitCooldown(kitName, kit.coolDown).also { targetUser.save() }
        kit.giveKit(targetPlayer)

        val messagePath = "Kit.Success." + (if (isSelf) "Self" else "Other")
        commandSender.commandMsg(messagePath) {
            target(targetPlayer.name)
            postModifier { it.replace("<KIT>", kit.displayName) }
        }
    }
}
